pkg/codegen/convert/plugin_mapper.go

```diff
@@ -173,12 +173,87 @@ func NewPluginMapper(ws Workspace,
 	}, nil
 }
 
+// getMappingForPlugin calls GetMapping on the given plugin and returns it's result. This currently takes the
+// target provider name as an argument so that if we don't find the mapping for "terraform" we can try again
+// for "tf".
+func (l *pluginMapper) getMappingForPlugin(targetProvider string, pluginSpec mapperPluginSpec) ([]byte, string, error) {
+	providerPlugin, err := l.providerFactory(pluginSpec.name, &pluginSpec.version)
+	if err != nil {
+		// We should maybe be lenient here and ignore errors but for now assume it's better to fail out on
+		// things like providers failing to start.
+		return nil, "", fmt.Errorf("could not create provider '%s': %w", pluginSpec.name, err)
+	}
+	contract.IgnoreClose(providerPlugin)
+
+	data, mappedProvider, err := providerPlugin.GetMapping(l.conversionKey)
+	if err != nil {
+		// This was an error calling GetMapping, not just that GetMapping returned a nil result. It's fine
+		// for GetMapping to return (nil, nil).
+		return nil, "", fmt.Errorf("could not get mapping for provider '%s': %w", pluginSpec.name, err)
+	}
+	// A provider returns empty if it didn't have a mapping
+	if mappedProvider != "" && len(data) != 0 {
+		// If this was the provider we we're looking for we can now return it
+		if mappedProvider == targetProvider {
+			return data, mappedProvider, nil
+		}
+	}
+	// TODO: Temporary hack to work around the fact that most of the plugins return a mapping for "tf" but
+	// not "terraform" but they're the same thing.
+	if l.conversionKey == "terraform" {
+		// Copy-pasta of the above _but_ we'll delete this whole if block once the plugins have had a
+		// chance to update.
+		data, mappedProvider, err = providerPlugin.GetMapping("tf")
+		if err != nil {
+			// This was an error calling GetMapping, not just that GetMapping returned a nil result. It's
+			// fine for GetMapping to return (nil, nil).
+			return nil, "", fmt.Errorf("could not get mapping for provider '%s': %w", pluginSpec.name, err)
+		}
+		// A provider returns empty if it didn't have a mapping
+		if mappedProvider != "" && len(data) != 0 {
+			// If this was the provider we we're looking for we can now return it
+			if mappedProvider == targetProvider {
+				return data, mappedProvider, nil
+			}
+		}
+	}
+
+	return data, mappedProvider, err
+}
+
 func (l *pluginMapper) GetMapping(provider string) ([]byte, error) {
 	// Do we already have an entry for this provider, if so use it
 	if entry, has := l.entries[provider]; has {
 		return entry, nil
 	}
 
+	// No entry yet, we're going to _try_ to get the plugin that matches the provider name as generally these
+	// will match up and it saves us doing a linear search through all plugins.
+	for i, pluginSpec := range l.plugins {
+		if pluginSpec.name == tokens.Package(provider) {
+			// Pop this out the list and try and get its mapping
+			l.plugins[i] = l.plugins[len(l.plugins)-1]
+			l.plugins = l.plugins[:len(l.plugins)-1]
```

Comment:
+1. We should not modify later items in the slice while ranging through it.

Looking at the rest of this code, though, it looks like what we're doing is:

- search through slice until we find a match
- if some conditions meet for that match, return it
- Otherwise break out of the loop

Which explains why modifying the slice while iterating didn't act weirdly.

I think this could be changed up a bit to address this concern:

- search through the slice to find a match
- record index and break
- rest of the loop body

```go
matchIdx := -1 // for invalid
for i, spec := range l.plugins {
  if spec.name == ... {
    matchIdx = i
    break
  }
}

if matchIdx >= 0 {
  l.plugins[matchIdx], l.plugins = l.plugins[len(l.plugins)-1], l.plugins[:len(l.plugins)-1]
  // ... rest of this code
}
```

Reply:
Yup if we mutate the list we stopped the iteration, but happy to rejig to a more explicit search then remove.